HomeAssistant/Authentication/AuthenticationController.swift

```diff
@@ -0,0 +1,97 @@
+//
+//  AuthenticationController.swift
+//  HomeAssistant
+//
+//  Created by Stephan Vanterpool on 8/11/18.
+//  Copyright © 2018 Robbie Trencheny. All rights reserved.
+//
+
+import Foundation
+import PromiseKit
+import SafariServices
+let kAuthenticationPath = "/auth/authorize?response_type=code&client_id=https%3A//blackgold9.github.io"
+    + "/home-assistant-iOS&redirect_uri=homeassistant%3A%2F%2Fauth-callback"
+
+class AuthenticationController: NSObject, SFSafariViewControllerDelegate {
+    enum AuthenticationControllerError: Error {
+        case invalidURL
+        case userCancelled
+    }
+
+    private var promiseResolver: Resolver<String>?
+    private var authenticationObserver: NSObjectProtocol?
+    private var authenticationViewController: SFSafariViewController?
+    var presentAuthenticationViewController: ((UIViewController) -> Void)?
+
+    override init() {
+        super.init()
+        self.configureAuthenticationObserver()
+    }
+
+    func authenticateWithBrowser(at baseURL: URL) -> Promise<String> {
+        return Promise { (resolver: Resolver<String>) in
+            self.promiseResolver = resolver
+            var components = URLComponents(url: baseURL, resolvingAgainstBaseURL: false)
+            components?.path = "/auth/authorize"
+            let responseTypeQuery = URLQueryItem(name: "response_type", value: "code")
+            let clientIDQuery = URLQueryItem(name: "client_id", value: "https://blackgold9.github.io/home-assistant-iOS")
+            let redirectQuery = URLQueryItem(name: "redirect_uri", value: "homeassistant://auth-callback")
+            components?.queryItems = [responseTypeQuery, clientIDQuery, redirectQuery]
+            if let newURL = try components?.asURL() {
+                let safariVC = SFSafariViewController(url: newURL)
+                if #available(iOS 11.0, *) {
+                    safariVC.dismissButtonStyle = .cancel
+                } else {
+                    // Fallback on earlier versions
```

Comment:
So what about iOS 10?

Reply:
It says done. Figured that was decent enough of a graceful fallback.